src/PerfCounters.cc

```diff
@@ -353,169 +249,14 @@ static void do_branches() {
-    int raw_fd = fd;
-    asm volatile(
-#if defined(__x86_64__)
-        "mov %[_SYS_ioctl], %%rax;"
-        "mov %[raw_fd], %%edi;"
-        "xor %%rdx, %%rdx;"
-        "mov %[_PERF_EVENT_IOC_ENABLE], %%rsi;"
-        "syscall;"
-        "cmp $-4095, %%rax;"
-        "jae 2f;"
-        "mov %[_SYS_ioctl], %%rax;"
-        "mov %[_PERF_EVENT_IOC_RESET], %%rsi;"
-        "syscall;"
-        // From this point on all conditional branches count!
-        "cmp $-4095, %%rax;"
-        "jae 2f;"
-        // Reset the counter period to the desired value.
-        "mov %[_SYS_ioctl], %%rax;"
-        "mov %[_PERF_EVENT_IOC_PERIOD], %%rsi;"
-        "mov %[period], %%rdx;"
-        "syscall;"
-        "cmp $-4095, %%rax;"
-        "jae 2f;"
-        "mov %[_iterations], %%rax;"
-        "1: dec %%rax;"
-        // Multiply by 7.
-        "mov %[accumulator], %%edx;"
-        "shl $3, %[accumulator];"
-        "sub %%edx, %[accumulator];"
-        // Add 2.
-        "add $2, %[accumulator];"
-        // Mask off bits.
-        "and $0xffffff, %[accumulator];"
-        // And loop.
-        "test %%rax, %%rax;"
-        "jnz 1b;"
-        "mov %[_PERF_EVENT_IOC_DISABLE], %%rsi;"
-        "mov %[_SYS_ioctl], %%rax;"
-        "xor %%rdx, %%rdx;"
-        // We didn't touch rdi.
-        "syscall;"
-        "cmp $-4095, %%rax;"
-        "jae 2f;"
-        "movl $0, %[count];"
-        "2: nop;"
-#elif defined(__i386__)
-        "mov %[_SYS_ioctl], %%eax;"
-        "mov %[raw_fd], %%ebx;"
-        "xor %%edx, %%edx;"
-        "mov %[_PERF_EVENT_IOC_ENABLE], %%ecx;"
-        "int $0x80;"
-        "cmp $-4095, %%eax;"
-        "jae 2f;"
-        "mov %[_SYS_ioctl], %%eax;"
-        "mov %[_PERF_EVENT_IOC_RESET], %%ecx;"
-        "int $0x80;"
-        // From this point on all conditional branches count!
-        "cmp $-4095, %%eax;"
-        "jae 2f;"
-        // Reset the counter period to the desired value.
-        "mov %[_SYS_ioctl], %%eax;"
-        "mov %[_PERF_EVENT_IOC_PERIOD], %%ecx;"
-        "mov %[period], %%edx;"
-        "int $0x80;"
-        "cmp $-4095, %%eax;"
-        "jae 2f;"
-        "mov %[_iterations], %%eax;"
-        "1: dec %%eax;"
-        // Multiply by 7.
-        "mov %[accumulator], %%edx;"
-        "shll $3, %[accumulator];"
-        "sub %%edx, %[accumulator];"
-        // Add 2.
-        "add $2, %[accumulator];"
-        // Mask off bits.
-        "andl $0xffffff, %[accumulator];"
-        // And loop.
-        "test %%eax, %%eax;"
-        "jnz 1b;"
-        "mov %[_PERF_EVENT_IOC_DISABLE], %%ecx;"
-        "mov %[_SYS_ioctl], %%eax;"
-        "xor %%edx, %%edx;"
-        // We didn't touch rdi.
-        "int $0x80;"
-        "cmp $-4095, %%eax;"
-        "jae 2f;"
-        "movl $0, %[count];"
-        "2: nop;"
+// Architecture specific detection code
+#if defined(__i386__) || defined(__x86_64__)
+#include "PerfCounters_x86.cc"
+#elif defined(__aarch64__)
+#include "PerfCounters_aarch64.cc"
```

Comment:
I think we should name these `.h` since they're not independent translation units. (E.g. it looks weird to me to have static functions in a `.cc` file that aren't used there.)

Reply:
Sure, I can do that. We can probably also make them separate translation units with the helpers all being members of `PerfCounters` and conditionally build them depending on the architecture. I'm open to either.

Reply:
Either's fine with me. I think it's probably a little simpler just to make them `.h` files.

Reply:
Easy enough, done.